### PR TITLE
Updating FastjetJetProducer (and related producers) with fillDescriptions (version2)

### PR DESCRIPTION
**Pull request by alefisico**
(Sorry to open a new PR, but #18832 is old already and it was a pain just to update it).

Update needed for using grooming techniques at HLT (previously incompatible with confdb). A brief summary of this PR:

Update FastjetJetProducer, and the classes affected by this change, with fillDescriptions method.
as discussed with @rappoccio @ferencek @ahinzmann @nhanvtran @jdolen @gkasieczka

addOnTests and runTheMatrix tests passed, here are the log files:
- https://cernbox.cern.ch/index.php/s/MLhMmXP5mVCkmth
- https://cernbox.cern.ch/index.php/s/gnYTrFQB5uDk9Ae

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py [hunk 1]
```diff
@@ -163,6 +163,16 @@ def customiseFor18429(process):
 
      return process
 
+# Updating FastjetJetProducer (and related producers) with fillDescriptions
+def customiseFor18832(process):
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py]: Please change this to ....18975.

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by Martin-Grunewald** [comment]
@perrotta @slava77 
please have a look and sign

File: RecoJets/JetProducers/plugins/FastjetJetProducer.cc [hunk 2]
```diff
@@ -520,6 +455,97 @@ void FastjetJetProducer::runAlgorithm( edm::Event & iEvent, edm::EventSetup cons
...
+	descFastjetJetProducer.add<unsigned int>("maxProblematicHcalCells",	9999999 );
+	descFastjetJetProducer.add<unsigned int>("maxRecoveredEcalCells",	9999999 );
+	descFastjetJetProducer.add<unsigned int>("maxRecoveredHcalCells",	9999999 );
+	vector<double>  puCentersDefault;
+	descFastjetJetProducer.add<vector<double>>("puCenters", 	puCentersDefault);
+	///// From PileUpSubstractor
+	descFastjetJetProducer.add<double> ("puPtMin", 	10.);
+	descFastjetJetProducer.add<double> ("nSigmaPU", 	1.);
+	descFastjetJetProducer.add<double> ("radiusPU", 	0.5);
+	descFastjetJetProducer.add<bool> ("sumRecHits", 	false);
```
Comment (perrotta) [inline RecoJets/JetProducers/plugins/FastjetJetProducer.cc]: This does not come from PileUpSubtractor (in this PR)...

In CMSSW_9_2_0, "sumRecHits" is only read by three HiJetAlgos modules:
- MultipleAlgoIterator
- ​ParametrizedSubtractor
- ​ReflectedIterator

They all derive from PileUpSubtractor, but this "sumRecHits" parameter is only defined in them (together with "dropZeroTowers"). It is not needed in this FastJetProducer, anyhow.

You had issues with removing this parameter in HLT with the customization function in the previous incarnation of this PR: did you try the suggestion of Martin in https://github.com/cms-sw/cmssw/pull/18832#issuecomment-303444805? Didn't it work?

File: RecoJets/JetProducers/plugins/FastjetJetProducer.cc [hunk 3]
```diff
@@ -520,6 +455,97 @@ void FastjetJetProducer::runAlgorithm( edm::Event & iEvent, edm::EventSetup cons
...
+	descFastjetJetProducer.add<unsigned int>("nExclude", 	0 	);
+	descFastjetJetProducer.add<unsigned int>("maxBadEcalCells", 	9999999	);
+	descFastjetJetProducer.add<unsigned int>("maxBadHcalCells",	9999999 );
+	descFastjetJetProducer.add<unsigned int>("maxProblematicEcalCells",	9999999 );
+	descFastjetJetProducer.add<unsigned int>("maxProblematicHcalCells",	9999999 );
+	descFastjetJetProducer.add<unsigned int>("maxRecoveredEcalCells",	9999999 );
+	descFastjetJetProducer.add<unsigned int>("maxRecoveredHcalCells",	9999999 );
+	vector<double>  puCentersDefault;
+	descFastjetJetProducer.add<vector<double>>("puCenters", 	puCentersDefault);
+	///// From PileUpSubstractor
```
Comment (perrotta) [inline RecoJets/JetProducers/plugins/FastjetJetProducer.cc]: PileUpSubtractor (just in case...)

**Review by perrotta — COMMENTED** [review]
As such the code works, and there is no effect on the jenkins outputs, as it should. 
There are two remaining issues I would like (at least) commented about:
- first, the sumRecHits parameter;
- then, I see that there is a lot of code duplication in the fillDescriptions() methods of CATopJetProducer : FastjetJetProducer : VirtualJetProducer. Is there any way to avoid it, in order to have a safer and better maintainable code? (I may think something like some shareable parameter set... but I don't know how it may be integrated and work, then I will not insist on it if you don't have a solution at hand)

**Comment by makortel** [comment]
> then, I see that there is a lot of code duplication in the fillDescriptions() methods of CATopJetProducer : FastjetJetProducer : VirtualJetProducer. Is there any way to avoid it, in order to have a safer and better maintainable code? (I may think something like some shareable parameter set... but I don't know how it may be integrated and work, then I will not insist on it if you don't have a solution at hand)

One possibilit would be to make a function taking `edm::ParameterSetDescription&` argument and adding the common parameters, and then call this function from the individual `fillDescriptions()` methods.

**Comment by fwyzard** [comment]
> One possibility would be to make a function taking an `edm::ParameterSetDescription&` argument and adding the common parameters, and then call this function from the individual fillDescriptions() methods.

`PoolSource` does [something similar](https://github.com/cms-sw/cmssw/blob/master/IOPool/Input/src/PoolSource.cc#L304) with its base class `InputSource`.

**Comment by fwyzard** [comment]
I would suggest to address the `sumRecHits` ASAP, and leave the restructuring of the `fillDescriptions` shared across multiple classes to a follow up PR.

@perrotta , would this be OK with you ?

File: RecoJets/JetProducers/plugins/FastjetJetProducer.cc (hunk 2 above)
Comment (alefisico) [inline RecoJets/JetProducers/plugins/FastjetJetProducer.cc]: @perrotta unfortunately the "sumRecHits" is needed in the FastjetJetproducer. The only way to get away from the problems here  https://github.com/cms-sw/cmssw/pull/18832#issuecomment-303207499 is to hardcode in the fillDescriptions. Unless you have something else to recommend.

**Comment by perrotta** [comment]
I think that restructuring the fillDescription methods can even ne  
more important. There are other PR queued which define new classes  
deriving from these ones: if left as such, also those classes will  
require intensive copy/paste, and the next update will have to ne  
propagated in several differenti places.

Both you and Matti proposed a solution: choose either one and apply it here


Andrea Bocci <notifications@github.com> ha scritto:

File: RecoJets/JetProducers/plugins/FastjetJetProducer.cc (hunk 3 above)
Comment (alefisico) [inline RecoJets/JetProducers/plugins/FastjetJetProducer.cc]: this is a general problem that I found everywhere. I would assume that if I define the fillDescriptions in, for instance, PileUpSubstractor and FasjetJetProducer uses PileUpSubstractor, then I would not need to add it in FastjetJetProducer but this is not the case. 
If I remove these lines from the PileUpSubstractor, then the runTheMatrix and addOnTests.py test failed with PileUpSubstractor related variables. Or did I forget to include something else?

**Comment by perrotta** [comment]
(I'm in train, I can't easily look at the code)

I couldn't find where it is needed... Anyhow, it si fine with me to  
let it there: at worst, there will ne a useless extra line of code  
that can be removed in future.

Alejandro Gomez Espinosa <notifications@github.com> ha scritto:

**Comment by fwyzard** [comment]
> Both you and Matti proposed a solution: choose either one and apply it here

Mhm, may I suggest that the Jets/MET RECO developers take care of the reorganisation - possibly after this PR is merged (and before any other related PR, but that is up to you to decide) ?

**Comment by rappoccio** [comment]
I will take care of the reorganization after this is done. Let's go with the simplest solution for now and we will make this more robust.

**Comment by perrotta** [comment]
@rappoccio @fwyzard @alefisico 

We missed the 921 release by now. I think the reorganization is important, also because there are other PRs waiting in the queue that will depend on it (see #18883): we cannot ask them to modify that PR once  when this one is merged for the first time, and once again after the reorganization.

Unless it is really urgent having this one merged before the we (please explain why if so) I would suggest updating it directly, merge it, and being done as such with it.

Please let me know if you see possible drawbacks in this plan.

**Comment by fwyzard** [comment]
Hi @perrotta ,
we need the changes in a release roughly by the end of next week, in order to make them available in the HLT, and let the PAGs prepare the new triggers.

**Comment by perrotta** [comment]
Thank you Andrea.

There could be a 922 sometime next week: let keep that target for this  
PR, then. I think it can be meet rather easily: the expected updates  
are quite technical and will be quickly approved as soon as the code  
runs and the filldescription generated configs actually contain what  
expected.

Andrea Bocci <notifications@github.com> ha scritto:

**Comment by alefisico** [comment]
@fwyzard @perrotta @rappoccio sorry we missed this deadline. I was at Fermilab for the B2G workshop and I have been with internet problems all week. 
Just to be sure, what exactly will be needed for this PR to get accepted? I'll try to work on this asap.

**Comment by perrotta** [comment]
Thank you Alejandro.

About "what exactly will be needed for this PR to get accepted", please refer to https://github.com/cms-sw/cmssw/pull/18975#pullrequestreview-41186850 and the following discussion, where the most important part is the removal of duplicated code where parameters are added. 

Please arrange with Sal who in this thread was proposing himself for taking care of the reorganization.

It would be great if you can have it done by Monday at latest

**Comment by alefisico** [comment]
Superseed by #19090

**Pull request closed without merging**